keps/sig-node/1769-memory-manager/README.md

```diff
@@ -0,0 +1,782 @@
+
+`SystemReserved` indicates the amount of memory which is pre-reserved, i.e. **guaranteed**, for the system, i.e., kernel, OS daemons, and core node components such as *kubelet* (*kubelet* deamons).
+
+#### Memory Maps at runtime (with examples)
+
+The figure below illustrates `Memory Maps` (for regular memory) during the node runtime. The figure shows the situation after a successful deployment of container _"A"_ and container _"B"_. Also, *Memory Manager* **guaranteed** (reserved) memory for the containers. In the example, the containers consume less memory than it was reserved (**guaranteed**), so for both containers their memory consumption remains below their memory limits (e.g. `cgroup` memory limit), and containers operate as normal.
+
+![alt text](./mem_tables_runtime.png)
+
+However, if **Out-of-Memory** (OOM) situation occurs, i.e., a particular memory limit is reached and crossed (e.g. by the containers) within _K8S_ node, the node triggers one of [the mechanisms](#kubernetes-nodes-memory-management-mechanisms-and-their-relation-to-the-memory-manager). Depending on what a particular limit is crossed (`cgroups` memory limit, `hard-eviction-treshold` memory limit, etc.), a suitable mechanism becomes triggered accordingly. As already noticed, such an OOM situation is handled by *kubelet* or system (OOM kernel killer) and not by *Memory Manager* itself. 
+
+### Simulation - how the Memory Manager works? (by examples)
+
+In this section, a number of slides is intended to illustrate the operation of the Memory Manager in various situations, i.e., to show how the pods in different QoS classes (Guaranteed, BestEffort/Burstable) are handled by the Memory Manager or other means, and how the disjoint groups of NUMA nodes are dynamically managed (created or removed) by the Memory Manager.  
+
+![alt text](./mm_simulation.jpg)
+
+#### Slide "Reject Pod2"
+
+The slide _"Reject Pod2"_ results from a standard scenario, i.e.:
+
+0. Both NUMA nodes are empty.
+1. Memory Manager finds out from which NUMA nodes a pod should allocate memory, i.e. *Pod2* must use both NUMA nodes. 
+2. Memory Manager updates `cpuset.mems`. 
+3. Kernel uses its default [NUMA Memory Policy](https://www.systutorials.com/docs/linux/man/2-set_mempolicy/), i.e. "local".
+4. Memory is allocated "locally", i.e. from the NUMA node where CPUs are pinned. 
+5. As normal, CPU Manager pinned CPUs to the first NUMA node.
+6. Therefore, memory is consumed from the first NUMA node completely (10G). 
+7. And then, *Pod2* consumes memory from the second NUMA node (5G).
+
+This might be surprising that *Pod2* is rejected although there is still enough spare memory at the second NUMA node. Nevertheless, the behaviour is correct and complies with both designs, i.e., disjoint NUMA node groups design and Topology Manager design.  
```

Comment:
We are following the CPU manager concept.
- if the container resource request can be satisfied by the single NUMA node, it should be pinned to the single NUMA node(single-numa and restricted topology manager policies)
- to start it, the system should have at least one NUMA node that currently not involved in the multi NUMA node allocation

Reply:
I tried to make it clear, first crucial thing is to understand TM design (link to TM blog was provided), as follows:
> This might be surprising that *Pod2* is rejected although there is still enough spare memory for *Pod2* at the second (single) NUMA node. Nevertheless, the behaviour is correct and complies with the design. In fact, [Topology Manager design](https://kubernetes.io/blog/2020/04/01/kubernetes-1-18-feature-topoloy-manager-beta/) involves the physical resource availability check: 
> \> "NOTE: Setting of the Preferred field in this way is not based on the set of currently available resources. It is based on the ability to physically allocate the number of requested resources on some minimal set of NUMA nodes."
> 
> So, in our case, *Pod2* can be allocated physically to a single NUMA node, but because there is no spare single NUMA node as *Group1* is an inseparable unit of two NUMA nodes, *Pod2* becomes rejected with the assumption that there is a spare single NUMA node somewhere in the cluster to satisfy *Pod2*. 

Also, elaboration by @klueska on Slack: [Thread 1](https://kubernetes.slack.com/archives/C012XSGFZQE/p1589542354108700)